### ISSUE TITLE
Prisma 2 not working with Nestjs

### ISSUE THREAD

**Issue by LeoMartinDev** [prisma/prisma#5938]
## Bug description

Prisma 2 is not working using latest NestJS.

![image](https://user-images.githubusercontent.com/7238941/109616771-da3f5980-7b35-11eb-9091-8d4db8113ddf.png)

## How to reproduce

Basic NestJS and Prisma setup does not work (see https://docs.nestjs.com/recipes/prisma)

## Expected behavior

We should be able to use Prisma

## Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model DiscordAuth {
  id           Int      @id @default(autoincrement())
  discordId    String   @unique
  accessToken  String   @unique
  refreshToken String   @unique
  tokenType    String
  expiresAt    DateTime
  scope        String
  userId       Int      @unique
  user         User     @relation(fields: [userId], references: [id])
}

model User {
  id          Int          @id @default(autoincrement())
  username    String
  email       String       @unique
  discordAuth DiscordAuth?
}

```

## Environment & setup

- OS: Windows - WSL 2 Ubuntu
- Database: PostgreSQL
- Node.js version: v14.15.0
- Prisma version:

```
prisma               : 2.17.0
@prisma/client       : 2.17.0
Current platform     : debian-openssl-1.1.x
Query Engine         : query-engine 3c463ebd78b1d21d8fdacdd27899e280cf686223 (at node_modules/@prisma/engines/query-engine-debian-openssl-1.1.x)
Migration Engine     : migration-engine-cli 3c463ebd78b1d21d8fdacdd27899e280cf686223 (at node_modules/@prisma/engines/migration-engine-debian-openssl-1.1.x)
Introspection Engine : introspection-core 3c463ebd78b1d21d8fdacdd27899e280cf686223 (at node_modules/@prisma/engines/introspection-engine-debian-openssl-1.1.x)
Format Binary        : prisma-fmt 3c463ebd78b1d21d8fdacdd27899e280cf686223 (at node_modules/@prisma/engines/prisma-fmt-debian-openssl-1.1.x)
Studio               : 0.353.0
```

**Comment by pantharshit00** [comment]
I am unable to reproduce this after following the tutorial step by step:
![image](https://user-images.githubusercontent.com/22195362/109632616-3bceea80-7b6d-11eb-94f4-c962c0264441.png)


It correctly bootstrapped the server correctly. Please make sure you are following the tutorial step by step properly. 

I am going to close this. If you can still reproduce this even after a second try, please upload the code to github repository so that I can take a look.

**Comment by LeoMartinDev** [comment]
Hmm just tried again and it still does not work 🤔 
Reproduction repository : https://github.com/LeoMartinDev/nestjs-prisma-bug
Node v14

**Comment by LeoMartinDev** [comment]
Reproduction steps on node v14 & using npm :

```
nest new prisma-nestjs
cd prisma-nestjs
npm install prisma --save-dev
npx prisma init
npx prisma migrate dev --name init --preview-feature
npm install @prisma/client
npm run start```

**Comment by pantharshit00** [comment]
@LeoMartinDev 

I looked into your code and you have a bug in there. In `src/app.controller.ts`, there is no route path defined in the `@Get` decorator and also the return type of the function is wrong.

Replace the `app.controller.ts` code to the following:
```ts
import { Controller, Get } from '@nestjs/common';
import { User } from '@prisma/client';
import { AppService } from './app.service';
import { PrismaService } from './prisma.service';

@Controller()
export class AppController {
  constructor(private prisma: PrismaService) {}

  @Get('/')
  async getHello(): Promise<User> {
    return this.prisma.user.findUnique({
      where: { email: 'ok' },
    });
  }
}
```


Also, please add `PrismaService` to `providers` in `app.module.ts`, otherwise nest can't perform dependency injection with it.

**Comment by LeoMartinDev** [comment]
@pantharshit00 thanks for taking a look :)

`@Get()` parameter is optional and resolves as `/` (https://docs.nestjs.com/controllers) so it should not be a concern.
However, I added `PrismaService` to AppModule's providers.
Sadly, I still have the same issue.
I updated the repository

**Comment by pantharshit00** [comment]
@LeoMartinDev Ah ok, good to know. I don't work with nest very often so  What you are missing here is the return type of the `getHello` function. You are returning a `Promise<User>` from it but return type is void right now. So it is not able to compile that. 

You can import `User` from '@prisma/client`. So change `getHello()` to `getHello(): Promise<User>`

**Comment by LeoMartinDev** [comment]
@pantharshit00 actually, it's not required as Typescript infer return type. It should at least throw a type error but here my whole app is instantly crashing without log.
IMHO this issue is from @prisma/client 
![image](https://user-images.githubusercontent.com/7238941/109669474-68d1cc00-7b72-11eb-8ee0-50f5c8877081.png)

**Comment by pantharshit00** [comment]
I initially though the decorator requires the type declaration but I am wrong there.

Mark the `getHello` function as `async` to resolve this. I don't know why this is failing but marking this as `async` seems to fix the problem with the decorator.

**Comment by LeoMartinDev** [comment]
> I initially though the decorator requires the type declaration but I am wrong there.
> 
> Mark the `getHello` function as `async` to resolve this. I don't know why this is failing but marking this as `async` seems to fix the problem with the decorator.

It does not (see my screenshot). It's not related to decorators and does not seem to be related to NestJs

![image](https://user-images.githubusercontent.com/7238941/109694001-962a7400-7b8a-11eb-8e5f-4a542c90f987.png)

**Comment by LeoMartinDev** [comment]
I mean, you can easily see that the issue comes from Prisma, simply comment everything related to it in my reproduction repository and the server will start properly

**Comment by LeoMartinDev** [comment]
😄 
![image](https://user-images.githubusercontent.com/7238941/109696075-f0c4cf80-7b8c-11eb-9621-5a82d16754d4.png)

**Comment by pantharshit00** [comment]
@LeoMartinDev I took a proper look this time. Sorry for this long thread. 

This is a typescript crash which indicates a bug in typescript compiler. If you run `npx tsc`, you will get `Debug Failure` which I believe nest internally runs. 
![image](https://user-images.githubusercontent.com/22195362/109697198-ec5fdd00-7bb3-11eb-9558-238c517ef497.png)


Please report this to typescipt.

**Comment by pantharshit00** [comment]
This is not a prisma bug as your code is valid typescript. This is typescript compiler crash

**Comment by LeoMartinDev** [comment]
> This is not a prisma bug as your code is valid typescript. This is typescript compiler crash. I suspect something is failing in type inference part of the compiler as adding a type annotation fixes the crash

Whose fault is it then, the error is gone if I comment everything related to Prisma 🤔 

The issue is not only impacting me as we managed to reproduct the issue on a fresh repository and I found several cases of people experiencing the problem.
Is Prisma not going to dig at least a bit into this issue ?

**Comment by pantharshit00** [comment]
Hey, looks like TS 4.3 dev fixes this.

![image](https://user-images.githubusercontent.com/22195362/109701804-75c5de00-7bb9-11eb-8513-1af2d3153744.png)


Can you verify this?

**Comment by LeoMartinDev** [comment]
👏 It works ! Using `"typescript": "^4.3.0-dev.20210302"`
![image](https://user-images.githubusercontent.com/7238941/109702185-35079e00-7b94-11eb-94d8-0ea89a961c26.png)

It should be fine for development but idk about production environment...

Thanks for your time / help @pantharshit00  ! 🎉

**Comment by Saggitarie** [comment]
@LeoMartinDev 
Thanks for asking this question. I was also encountering the same issue at around the same time.
Glad that there is a workaround at least for now!

@pantharshit00 Thanks for the research on the error. TBH, I'm starting to really like using prisma with nestjs. 
This might/might not solve until the next stable release on Typescript 4.3. But we can make sure that it working, I think this issues should be reopened.

**Comment by V-ed** [comment]
Using Prisma `2.17.0` and Typescript `4.2.2` :

<table>
<tr>
	<td>Return type on functions explicitely set
	<td>Return type on <code>getAll()</code> function removed
<tr>
	<td>
<img src="https://user-images.githubusercontent.com/16652915/110169313-86976f00-7dc6-11eb-8012-06d534c41f65.png" />
	<td>
<img src="https://user-images.githubusercontent.com/16652915/110169728-1b01d180-7dc7-11eb-8b9b-af738f9cca86.png" />

</table>

In other words, I've fixed this on my side by explicitly tagging my functions with the return type expected.

Happy to see that it works correctly in Typescript `4.3`, but I'm currently using the eslint rule `"@typescript-eslint/explicit-function-return-type": ["error"]` so that I remember to add all return types and not have to deal with a `Debug Failure` again, and besides, it's best practice to to it too so why not hehe

Hopefully my 8 hours debugging time of yesterday wasn't for nothing :)

**Comment by pantharshit00** [comment]
This is also fixed in typescript 4.2.3 so I will not reopen this now. TS 4.2.3 was released 3 days ago. I believe this the issue which got patched by the TS team: https://github.com/microsoft/TypeScript/issues/43011
